File: legacy/react-native/utils/proration.ts
/**
 * Days a member was present in a cycle, inclusive of both endpoints.
 * Returns 0 if move-out is before cycle start.
 */
export function daysPresent(
  cycleStartIso: string,
  cycleEndIso: string,
  moveOutIso: string,
): number {
  const start = parseIso(cycleStartIso);
  const end = parseIso(cycleEndIso);
  const move = parseIso(moveOutIso);
  if (move < start) return 0;
  const last = move > end ? end : move;
  const ms = last.getTime() - start.getTime();
  return Math.floor(ms / (1000 * 60 * 60 * 24)) + 1;
}

export function cycleTotalDays(cycleStartIso: string, cycleEndIso: string): number {
  const start = parseIso(cycleStartIso);
  const end = parseIso(cycleEndIso);
  const ms = end.getTime() - start.getTime();
  return Math.floor(ms / (1000 * 60 * 60 * 24)) + 1;
}

/**
 * Prorate a full amount by days-present fraction. Rounded to cents.
 */
export function prorateAmount(
  fullAmount: number,
  cycleStartIso: string,
  cycleEndIso: string,
  moveOutIso: string,
): number {
  if (!isValidIso(cycleStartIso) || !isValidIso(cycleEndIso) || !isValidIso(moveOutIso)) {
    return 0;
  }
  const present = daysPresent(cycleStartIso, cycleEndIso, moveOutIso);
  const total = cycleTotalDays(cycleStartIso, cycleEndIso);
  if (total <= 0) return 0;
  return parseFloat(((fullAmount * present) / total).toFixed(2));
}

const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;

function isValidIso(iso: string): boolean {
  return ISO_DATE.test(iso);
}

function parseIso(iso: string): Date {
  if (!ISO_DATE.test(iso)) return new Date(0);
  const parts = iso.split('-').map(Number);
  const [y, m, d] = parts as [number, number, number];
  return new Date(Date.UTC(y, m - 1, d));
}

File: HomesplitIOS/Domain/Proration/Proration.swift
import Foundation

enum Proration {
    private static let isoDatePattern = #"^\d{4}-\d{2}-\d{2}$"#
    private static let secondsPerDay: TimeInterval = 86_400

    /// Days a member was present in a cycle, inclusive of both endpoints.
    /// Returns 0 if move-out is before cycle start.
    static func daysPresent(
        cycleStartIso: String,
        cycleEndIso: String,
        moveOutIso: String
    ) -> Int {
        let start = parseIso(cycleStartIso)
        let end = parseIso(cycleEndIso)
        let move = parseIso(moveOutIso)
        if move < start { return 0 }
        let last = move > end ? end : move
        let seconds = last.timeIntervalSince(start)
        return Int(floor(seconds / secondsPerDay)) + 1
    }

    static func cycleTotalDays(cycleStartIso: String, cycleEndIso: String) -> Int {
        let start = parseIso(cycleStartIso)
        let end = parseIso(cycleEndIso)
        let seconds = end.timeIntervalSince(start)
        return Int(floor(seconds / secondsPerDay)) + 1
    }

    /// Prorate a full amount by the days-present fraction. Rounded to cents.
    static func prorateAmount(
        fullAmount: Decimal,
        cycleStartIso: String,
        cycleEndIso: String,
        moveOutIso: String
    ) -> Decimal {
        guard isValidIso(cycleStartIso),
              isValidIso(cycleEndIso),
              isValidIso(moveOutIso)
        else { return 0 }

        let present = daysPresent(
            cycleStartIso: cycleStartIso,
            cycleEndIso: cycleEndIso,
            moveOutIso: moveOutIso
        )
        let total = cycleTotalDays(cycleStartIso: cycleStartIso, cycleEndIso: cycleEndIso)
        if total <= 0 { return 0 }
        return (fullAmount * Decimal(present) / Decimal(total))
            .roundedHalfAwayFromZero(scale: 2)
    }

    static func isValidIso(_ iso: String) -> Bool {
        iso.range(of: isoDatePattern, options: .regularExpression) != nil
    }

    static func parseIso(_ iso: String) -> Date {
        guard isValidIso(iso) else { return Date(timeIntervalSince1970: 0) }
        let parts = iso.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3 else { return Date(timeIntervalSince1970: 0) }
        var components = DateComponents()
        components.year = parts[0]
        components.month = parts[1]
        components.day = parts[2]
        components.timeZone = TimeZone(identifier: "UTC")
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        return calendar.date(from: components) ?? Date(timeIntervalSince1970: 0)
    }
}
